runtime/protocol/http-client-engines/http-client-engine-okhttp/jvm/src/aws/smithy/kotlin/runtime/http/engine/okhttp/OkHttpEngine.kt

```diff
@@ -51,7 +55,14 @@ public class OkHttpEngine(
         val engineResponse = mapOkHttpExceptions { engineCall.executeAsync() }
 
         callContext.job.invokeOnCompletion {
-            engineResponse.body.close()
+            // FIXME: https://github.com/awslabs/smithy-kotlin/issues/935
+            val closeResult = runCatching { engineResponse.body.close() }
+
+            if (closeResult.exceptionOrNull() is IllegalStateException) {
+                engineCall.cancel()
+                runBlocking { delay(1.nanoseconds) }
```

Comment:
[invokeOnCompletion](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-job/invoke-on-completion.html) shouldn't block according to the documentation.

Reply:
Hmm... this might be difficult to work around then. Without this `delay` the IllegalStateException is still thrown on `.close()`, even when running `engineCall.cancel()` beforehand.

Reply:
Yeah I _think_ the way it might work if we can get the following in place:
 1. tie the background coroutine to the call scope
 2. detect when we need to call `cancel` on the call (will require making a call an interface or something so that we can abstract this per/engine)
 3. in `HttpCall.complete` call cancel when necessary _before_ calling `join()` on the call context job. 

This would make the existing completion handler that closes the response body only be invoked _after_ the corouting doing the `read()` has exited.

This is obviously a bigger change.